services/base.py

```diff
@@ -380,29 +396,77 @@ def _fetchMemory(self, key):
       self._MEMORY = []
     self._MEMORY_KEY = h
 
-  def memoryRemember(self, itemId, keywords=None):
+  def _differentThanLastHistory(self, keywordindex, imageIndex):
+    if len(self._HISTORY) == 0:
+      return True
+    if keywordindex == self._HISTORY[-1][0] and imageIndex == self._HISTORY[-1][1]:
+      return False
+    return True
+
+  def memoryRemember(self, itemId, keywords=None, alwaysRemember=True):
+    logging.debug("Displaying new image")
+    logging.debug(self._NAME)
+    logging.debug("keyword: %d; index: %d" % (self.keywordIndex, self.imageIndex))
+
     self._fetchMemory(keywords)
     h = self.hashString(itemId)
-    if h in self._MEMORY:
-      return
-    self._MEMORY.append(h)
+    if h not in self._MEMORY:
+      self._MEMORY.append(h)
+    
+    # only remember if image has changed
+    rememberInHistory = alwaysRemember or self._differentThanLastHistory(self.keywordIndex, self.imageIndex)
+    if rememberInHistory:
+      self._HISTORY.append((self.keywordIndex, self.imageIndex))
+    self.imageIndex += 1
 
     if (len(self._MEMORY) % 20) == 0:
       logging.info('Interim saving of memory every 20 entries')
       with open(os.path.join(self._DIR_MEMORY, '%s.json' % self._MEMORY_KEY), 'w') as f:
         json.dump(self._MEMORY, f)
 
+    return rememberInHistory
+
   def memorySeen(self, itemId, keywords=None):
     self._fetchMemory(keywords)
     h = self.hashString(itemId)
     return h in self._MEMORY
 
-  def memoryForget(self, keywords=None):
+  def memoryForgetLast(self, keywords=None):
+    self._fetchMemory(keywords)
+    if len(self._MEMORY) != 0:
+      self._MEMORY.pop()
+    if len(self._HISTORY) == 0:
+      logging.warning("Trying to forget a single memory, but 'self._HISTORY' is empty. This should have never happened!")
+    else:
+      self.keywordIndex, self.imageIndex = self._HISTORY.pop()
+
+  def memoryForget(self, keywords=None, forgetHistory=False):
     self._fetchMemory(keywords)
     n = os.path.join(self._DIR_MEMORY, '%s.json' % self._MEMORY_KEY)
     if os.path.exists(n):
       os.unlink(n)
     self._MEMORY = []
+    if forgetHistory:
+      self._HISTORY = []
+
+  def getPreferredImageSize(self, imageSize, displaySize):
+    # Calculate the size we need to avoid black borders
```

Comment:
It's a little hard to explain, but I tried my best with the comment:
https://github.com/mrworf/photoframe/blob/c71f40d6a673ea9622298bd8ae632fe8a52633f6/services/base.py#L520-L525

It basically returns the smallest possible size, so that no upscaling needs to be done during _processing_ (regardless of scaling option).